account-kit/logging/src/index.ts

```diff
@@ -1,15 +1,53 @@
 import { createClientLogger } from "./client.js";
 import { createServerLogger } from "./server.js";
-import type { EventLogger, EventsSchema } from "./types";
+import type { EventLogger, EventsSchema, LoggerContext } from "./types";
 
 export type * from "./types.js";
 
-export function createLogger<
-  Schema extends EventsSchema
->(): EventLogger<Schema> {
-  if (typeof window === "undefined") {
-    return createServerLogger();
-  }
+export function createLogger<Schema extends EventsSchema = []>(
+  context: LoggerContext
+): EventLogger<Schema>;
 
-  return createClientLogger();
+export function createLogger(context: LoggerContext): EventLogger {
+  const innerLogger =
+    typeof window === "undefined"
+      ? createServerLogger(context)
+      : createClientLogger(context);
+
+  const logger: EventLogger = {
+    ...innerLogger,
+    profiled<TArgs extends any[], TRet>(
+      name: string,
+      func: (...args: TArgs) => TRet
+    ): (...args: TArgs) => TRet {
+      return function (...args: TArgs): TRet {
+        const start = Date.now();
+        const result = func(...args);
```

Comment:
Should do `func.apply(this, args)` instead of `func(...args)`, otherwise you lose the value of `this`.

Reply:
should `this` here be the `func`?

I thought if we did `func.apply(this...` the `this` here is actually object we're returning which contains `profiled` and `trackEvent` in it?

Reply:
It would do that if we were returning an arrow function e.g.
```ts
return (...args: TArgs): TRet => { /* ... */ }
```
because arrow functions capture the surrounding value of `this` which is more often what you want. But since we're returning a function declared with the `function` keyword, `this` will refer to the object that the returned function is ultimately "called on": that is, whatever value appears to the left of the dot when this function is called with method call syntax. So for example, in
```ts
const sender = {
    sendUserOperation: ReactLogger.profiled(
      "sendUserOperation",
      sendUserOperation
    ),
};

sender.sendUserOperation(op);
```
in the call to `sender.sendUserOperation(op)`, `this` will be `sender`.